### PR TITLE
Image, Pixels and Channel encoders, and decoders

### PR DESCRIPTION
**Pull request by chris-allan**
Updated version of #10 with all fields encoded and decoded. Expanded to include `Channel`.

### CODE REVIEW

File: omero_marshal/encode/encoders/channel.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    logical_channel_1.name = rstring('GFP/488')
+    logical_channel_1.ndFilter = rdouble(1.0)
+    logical_channel_1.pinHoleSize = LengthI(1.0, UnitsLength.NANOMETER)
+    logical_channel_1.pockelCellSetting = rint(0)
+    logical_channel_1.samplesPerPixel = rint(2)
+    logical_channel_1.contrastMethod = contrast_method
+    logical_channel_1.illumination = illumination
+    logical_channel_1.mode = acquisition_mode
+    logical_channel_1.photometricInterpretation = photometric_interpretation
+"""
```
Comment (will-moore) [inline omero_marshal/encode/encoders/channel.py]: Is this comment block meant to be left in?

File: omero_marshal/encode/encoders/image.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        _format = obj.format
+        if _format is not None and _format.isLoaded():
+            format_encoder = self.ctx.get_encoder(obj.format.__class__)
+            self.set_if_not_none(
+                v, 'omero:format', format_encoder.encode(obj.format)
+            )
+        if obj.isPixelsLoaded() and obj.sizeOfPixels() > 0:
+            pixels = obj.getPrimaryPixels()
+            pixels_encoder = self.ctx.get_encoder(pixels.__class__)
+            v['Pixels'] = pixels_encoder.encode(pixels)
```
Comment (chris-allan) [inline omero_marshal/encode/encoders/image.py]: The schema hasn't supported them for several years now, so nope. :)

File: omero_marshal/decode/decoders/channel.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        v = super(Channel201501Decoder, self).decode(data)
+        self.set_property(v, 'alpha', data.get('Alpha'))
+        self.set_property(v, 'blue', data.get('Blue'))
+        self.set_property(v, 'green', data.get('Green'))
+        self.set_property(v, 'red', data.get('Red'))
+        self.set_property(v, 'lookupTable', data.get('omero:lookupTable'))
+
+        logical_channel_id = data.get('omero:LogicalChannelId')
+        if logical_channel_id is None:
+            return v
```
Comment (will-moore) [inline omero_marshal/decode/decoders/channel.py]: If I'm creating a new Channel, to save to OMERO, I won't have a LogicalChannelId. So how do I create a new LogicalChannel in OMERO? Can we not create a new, unsaved LogicalChannel here?

**Comment by will-moore** [comment]
Tested image and channels marshalling of real OMERO data and looks good, but tricky to test this more fully on real data without writing a bunch of queries. I'll start working on this with the web_api soon.
Tests all pass for me locally.

Not specific to this PR but I was wondering how are we going to make things like rendering settings available to web users? (to provide similar functionality to current imgData json?)
Another general observation is that the 'omero:details' really adds to the verbosity of the json, E.g. mostly users aren't going to need this on Pixels['Type'] etc. Being able to exclude or normalise  'omero:details' would be great.

**Comment by chris-allan** [comment]
Agreed, @will-moore.

We've already got the normalisation requirement recorded in #22. Probably will have to wait for 0.5.0 or later as it will require a new API with a dynamic context and/or configuration properties. I've not given this any serious thought.

As far as rendering settings are concerned we probably just need to work with @jburel, @sbesson, @joshmoore and others on a preliminary model and marshal them as well. Not sure we have another option.

**Comment by sbesson** [comment]
Good to merge from my perspective as well. Discussed with @chris-allan who might try to also add support for `Well/WellSample` in order to complete the `PDI` and `SPW` graph before we cut a 0.5.0 release of omero-marshal.

**Pull request merged**